File: Java/0151-0200/189. Rotate Array.java
// https://leetcode.com/problems/rotate-array/

class Solution {
    // Function to reverse elements in the Array
    public void reverse(int[] nums, int start, int end){
        while(start < end){
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;end--;
        }
    }

    public void rotate(int[] nums, int k) {
        int size = nums.length;
        k %= size; // ensure k is in the range

        reverse(nums, 0, size-1); // Revere Entire Array
        reverse(nums, 0 , k-1); // Rrverse K elements
        reverse(nums, k, size-1); // Reverse remaining elements
    }
}

File: Python/0151-0200/189. Rotate Array.py
# https://leetcode.com/problems/rotate-array/

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        # Function to reverse elements in the Array
        def reverse(nums, start, end):
            while start < end :
                temp = nums[start]
                nums[start] = nums[end]
                nums[end] = temp
                start += 1
                end -= 1
        
        size = len(nums)
        k %= size # ensure k is in the range

        reverse(nums, 0, size-1) # Revere Entire Array
        reverse(nums, 0 , k-1) # Rrverse K elements
        reverse(nums, k, size-1) # Reverse remaining elements
